### PR TITLE
Automatically Size PerfEventArrays

### PR DESCRIPTION
**Pull request by dave-tucker**
This brings Aya into compliance with libbpf and other popular loaders.
PerfEventArray sizes should be set to 0 at compile time.
At load time, they are set to the number of online CPUs.

Fixes: #995

### CODE REVIEW

File: aya/src/maps/bloom_filter.rs [hunk 1]
```diff
@@ -120,10 +120,8 @@ mod tests {
 
     #[test]
     fn test_try_from_wrong_map() {
-        let map = new_map(test_utils::new_obj_map::<u32>(
-            BPF_MAP_TYPE_PERF_EVENT_ARRAY,
-        ));
-        let map = Map::PerfEventArray(map);
+        let map = new_map(test_utils::new_obj_map::<u32>(BPF_MAP_TYPE_ARRAY));
```
Comment (dave-tucker) [inline aya/src/maps/bloom_filter.rs:123]: Note to reviewers: This is necessary to stop miri tripping over the PERF_EVENT_ARRAY logic and attempting to `open` a file to read number of online CPUs

File: aya/src/sys/bpf.rs [hunk 2]
```diff
@@ -46,6 +46,20 @@ pub(crate) fn bpf_create_map(
     u.map_type = def.map_type();
     u.key_size = def.key_size();
     u.value_size = def.value_size();
+    u.max_entries = def.max_entries();
+
+    // BPF_MAP_TYPE_PERF_EVENT_ARRAY's max_entries should be 0
+    // There are some (older) cases where this is a large constant
+    // Set to the number of online CPUs if it's 0
+    // Clamp to (at most) the number of online CPUs if it's larger than that
+    // Smaller values are left as is as this was required by some kernel selftests
```
Comment (vadorovsky) [inline aya/src/sys/bpf.rs:55]: This comment is a bit difficult to read. I also don't understand two points:

- "There are some (older) cases where this is a large constant" - I didn't get it at all. Could you elaborate a bit?
- "Smaller values are left as is as this was required by some kernel selftests" - I don't get that as well. My understanding is that we should allow values greater than 0 and lower than number of CPUs for loading programs not written in Aya (or using old versions of aya-ebpf and hard coding high values). Is my understanding correct?

From what I was able to understand, I would rephrase it as:

```suggestion
    // BPF_MAP_TYPE_PERF_EVENT_ARRAY's max_entries should not exceed the number of
    // CPUs.
    //
    // By default, the newest versions of Aya, libbpf and cilium/ebpf define `max_entries` of
    // `PerfEventArray` as `0`, with an intention to get it replaced with a correct value
    // by the loader.
    //
    // We allow custom values (potentially coming either from older versions of aya-ebpf or
    // programs written in C) as long as they don't exceed the number of CPUs.
    //
    // Otherwise, when the value is `0` or too large, we set it to the number of CPUs.
```

**Review by vadorovsky — CHANGES_REQUESTED** [review]
The code looks good, but I'm nitpicking the code comment. :stuck_out_tongue: 

Also, from what I quickly checked, we don't have an integration test with `PerfEventArray`. It would be nice to add it (maybe as a separate PR).

File: aya/src/sys/bpf.rs (hunk 2 above)
Comment (dave-tucker) [reply aya/src/sys/bpf.rs]: Good call. Incorporated your comment and re-pushed.

File: aya/src/maps/bloom_filter.rs (hunk 1 above)
Comment (tamird) [reply aya/src/maps/bloom_filter.rs:123]: That sounds like it should be a code comment.

**Review by vadorovsky — APPROVED** [review]
Awesome!

File: aya/src/maps/bloom_filter.rs (hunk 1 above)
Comment (vadorovsky) [reply aya/src/maps/bloom_filter.rs:123]: I agree. Let's make it a code comment before merging.

**Comment by tamird** [comment]
FYI this broke the book build: https://github.com/aya-rs/book/actions/runs/10156051789
